File: crear.js
const { MongoClient } = require('mongodb');
const inquirer = require('inquirer');

// Conexión a MongoDB (Actualiza con tu URI)
const mongoURI = 'mongodb+srv://dia21066:CMSOw7YTLnPTsdEp@cluster0.a4fng.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0';

const connectToMongoDB = async () => {
  const client = new MongoClient(mongoURI, { useNewUrlParser: true, useUnifiedTopology: true });
  await client.connect();
  console.log("Conectado a MongoDB");
  return client;
};

// Crear Usuario
const crearUsuario = async () => {
  const { nombre, correo, direccion, telefono } = await inquirer.prompt([
    { type: 'input', name: 'nombre', message: 'Ingrese el nombre del usuario:' },
    { type: 'input', name: 'correo', message: 'Ingrese el correo del usuario:' },
    { type: 'input', name: 'direccion', message: 'Ingrese la dirección del usuario:' },
    { type: 'input', name: 'telefono', message: 'Ingrese el teléfono del usuario:' },
  ]);

  const client = await connectToMongoDB();
  const db = client.db('Proyecto_2');
  const usuariosCollection = db.collection('Usuarios');

  const nuevoUsuario = { nombre, correo, direccion, telefono };
  await usuariosCollection.insertOne(nuevoUsuario);

  console.log('Usuario creado con éxito');
  await client.close();
};

// Crear Restaurante
const crearRestaurante = async () => {
  const { nombre, direccion, categoria } = await inquirer.prompt([
    { type: 'input', name: 'nombre', message: 'Ingrese el nombre del restaurante:' },
    { type: 'input', name: 'direccion', message: 'Ingrese la dirección del restaurante:' },
    { type: 'input', name: 'categoria', message: 'Ingrese la categoría del restaurante:' },
  ]);

  const client = await connectToMongoDB();
  const db = client.db('Proyecto_2');
  const restaurantesCollection = db.collection('Restaurantes');

  const nuevoRestaurante = { nombre, direccion, categoria };
  await restaurantesCollection.insertOne(nuevoRestaurante);

  console.log('Restaurante creado con éxito');
  await client.close();
};

// Crear Artículo de Menú
const crearArticuloDeMenu = async () => {
  const { nombre, descripcion, precio, restauranteid, categoria } = await inquirer.prompt([
    { type: 'input', name: 'nombre', message: 'Ingrese el nombre del artículo:' },
    { type: 'input', name: 'descripcion', message: 'Ingrese la descripción del artículo:' },
    { type: 'input', name: 'precio', message: 'Ingrese el precio del artículo:' },
    { type: 'input', name: 'restauranteid', message: 'Ingrese el ID del restaurante:' },
    { type: 'input', name: 'categoria', message: 'Ingrese la categoría del artículo:' },
  ]);

  const client = await connectToMongoDB();
  const db = client.db('Proyecto_2');
  const articulosCollection = db.collection('Articulos_Menu');

  const nuevoArticulo = {
    nombre,
    descripcion,
    precio: parseFloat(precio),
    restauranteid,
    categoria,
  };
  await articulosCollection.insertOne(nuevoArticulo);

  console.log('Artículo de menú creado con éxito');
  await client.close();
};

// Crear Orden
const crearOrden = async () => {
  const { restauranteid, usuarioid, articulos } = await inquirer.prompt([
    { type: 'input', name: 'restauranteid', message: 'Ingrese el ID del restaurante:' },
    { type: 'input', name: 'usuarioid', message: 'Ingrese el ID del usuario:' },
    { type: 'input', name: 'articulos', message: 'Ingrese los artículos (JSON):' },
  ]);

  const articulosArray = JSON.parse(articulos);
  const client = await connectToMongoDB();
  const db = client.db('Proyecto_2');
  const ordenesCollection = db.collection('Ordenes');

  const nuevaOrden = {
    restauranteid,
    usuarioid,
    datosMenu: articulosArray,
    total: articulosArray.reduce((sum, articulo) => sum + articulo.precio, 0),
    estado: 'Pendiente',
    fechaOrden: new Date(),
  };
  await ordenesCollection.insertOne(nuevaOrden);

  console.log('Orden creada con éxito');
  await client.close();
};

// Crear Reseña
const crearResena = async () => {
  const { usuarioid, restauranteid, ordenid, calificacion, comentario } = await inquirer.prompt([
    { type: 'input', name: 'usuarioid', message: 'Ingrese el ID del usuario:' },
    { type: 'input', name: 'restauranteid', message: 'Ingrese el ID del restaurante:' },
    { type: 'input', name: 'ordenid', message: 'Ingrese el ID de la orden:' },
    { type: 'input', name: 'calificacion', message: 'Ingrese la calificación (1-5):' },
    { type: 'input', name: 'comentario', message: 'Ingrese el comentario:' },
  ]);

  const client = await connectToMongoDB();
  const db = client.db('Proyecto_2');
  const resenasCollection = db.collection('Resenas');

  const nuevaResena = {
    usuarioid,
    restauranteid,
    ordenid,
    calificacion: parseFloat(calificacion),
    comentario,
    fechaResena: new Date(),
  };
  await resenasCollection.insertOne(nuevaResena);

  console.log('Reseña creada con éxito');
  await client.close();
};

//exportar funciones
module.exports = {
  crearUsuario,
  crearRestaurante,
  crearArticuloDeMenu,
  crearOrden,
  crearResena,
};

File: crear.py
from flask import jsonify, request
from pymongo import MongoClient
from bson import ObjectId

# Conexión a MongoDB
def connect_to_mongo():
    mongo_uri = 'mongodb+srv://dia21066:CMSOw7YTLnPTsdEp@cluster0.a4fng.mongodb.net/?retryWrites=true&w=majority'
    client = MongoClient(mongo_uri)
    db = client['Proyecto_2']  # Reemplaza con el nombre de tu base de datos
    return db

# Crear un usuario (Pedir campos por separado)
def crear_usuario(request, db):
    try:
        # Pedir los campos por separado
        nombre = request.json.get("nombre")
        correo = request.json.get("correo")
        direccion = request.json.get("direccion")
        telefono = request.json.get("telefono")
        fecha_registro = request.json.get("fechaRegistro")
        fecha_nacimiento = request.json.get("fechaNacimiento")

        # Crear el objeto JSON
        usuario = {
            "nombre": nombre,
            "correo": correo,
            "direccion": direccion,
            "telefono": telefono,
            "fechaRegistro": fecha_registro,
            "fechaNacimiento": fecha_nacimiento
        }

        # Insertar el nuevo usuario en la colección 'usuarios'
        result = db.Usuarios.insert_one(usuario)

        # Responder con un mensaje de éxito
        return jsonify({"message": f"Usuario creado con éxito, ID: {str(result.inserted_id)}"}), 201
    except Exception as e:
        # Manejo de errores
        return jsonify({"message": f"Error al crear el usuario: {str(e)}"}), 500


# Crear un restaurante (Pedir campos por separado)
def crear_restaurante(request, db):
    try:
        # Pedir los campos por separado
        nombre_restaurante = request.json.get("nombre")
        descripcion = request.json.get("descripcion")
        categoria = request.json.get("categoria")
        calle = request.json.get("calle")
        ciudad = request.json.get("ciudad")
        telefono = request.json.get("telefono")
        longitud = request.json.get("longitud")
        latitud = request.json.get("latitud")
        fecha_registro = request.json.get("fechaRegistro")

        # Validar que no falten campos requeridos
        if not nombre_restaurante or not descripcion or not categoria:
            return jsonify({"message": "Faltan datos obligatorios: nombre, descripcion, categoria"}), 400
        
        if longitud is None or latitud is None:
            return jsonify({"message": "Las coordenadas no pueden ser nulas."}), 400

        # Crear el objeto JSON para el restaurante
        restaurante = {
            "nombre": nombre_restaurante,
            "descripcion": descripcion,
            "categoria": categoria,
            "direccion": {
                "calle": calle,
                "ciudad": ciudad
            },
            "telefono": telefono,
            "ubicacion": {
                "type": "Point",
                "coordinates": [longitud, latitud]  # Usando la longitud y latitud como array de coordenadas
            },
            "fechaRegistro": fecha_registro
        }

        # Insertar el nuevo restaurante en la colección 'restaurantes'
        result = db.Restaurantes.insert_one(restaurante)

        # Responder con un mensaje de éxito
        return jsonify({"message": f"Restaurante creado con éxito, ID: {str(result.inserted_id)}"}), 201
    except Exception as e:
        # Manejo de errores
        return jsonify({"message": f"Error al crear el restaurante: {str(e)}"}), 500


# Crear un artículo de menú (Pedir campos por separado)
def crear_articulo_de_menu(request, db):
    try:
        # Pedir los campos por separado
        nombre_articulo = request.json.get("nombre")
        precio = request.json.get("precio")
        descripcion = request.json.get("descripcion")
        disponible = request.json.get("disponibilidad")
        restaurante_id = request.json.get("restauranteId")
        categoria = request.json.get("categoria")

        # Crear el objeto JSON para el artículo de menú
        articulo = {
            "nombre": nombre_articulo,
            "precio": precio,
            "descripcion": descripcion,
            "disponibilidad": disponible,
            "restauranteId": restaurante_id,
            "categoria": categoria  
        }

        # Insertar el artículo de menú en la colección 'articulos_menu'
        result = db.Articulos_Menu.insert_one(articulo)

        # Responder con un mensaje de éxito
        return jsonify({"message": f"Artículo de menú creado con éxito, ID: {str(result.inserted_id)}"}), 201
    except Exception as e:
        # Manejo de errores
        return jsonify({"message": f"Error al crear el artículo de menú: {str(e)}"}), 500


# Crear una orden 
def crear_orden(request, db):
    try:
        # Pedir los campos por separado
        usuario_id = request.json.get("usuarioId")
        restaurante_id = request.json.get("restauranteId")
        items_ids = request.json.get("items")  # Esto debería ser una lista de ids de artículos
        estado = request.json.get("estado")
        fecha_orden = request.json.get("fechaOrden")

        # Obtener los datos del restaurante
        restaurante = db.Restaurantes.find_one({"_id": restaurante_id})
        if not restaurante:
            return jsonify({"message": "Restaurante no encontrado"}), 404

        # Obtener los datos del usuario
        usuario = db.Usuarios.find_one({"_id": ObjectId(usuario_id)})
        datos = []
        if usuario:
            items = {
                "usuarioId": usuario_id,
                "nombre": usuario.get("nombre"),
                "correo": usuario.get("correo")
            }
            datos.append(items)
        else:
            return jsonify({"message": "Usuario no encontrado"}), 404
        

        # Obtener los artículos y calcular el total
        total = 0
        datos_menu = []
        for item_id in items_ids:
            articulo = db.Articulos_Menu.find_one({"_id": item_id})
            if articulo:
                item = {
                    "articuloId": item_id,
                    "nombreArticulo": articulo.get("nombreArticulo"),
                    "precio": articulo.get("precio")
                }
                datos_menu.append(item)
                total += articulo.get("precio", 0)  # Acumular el precio al total
            else:
                return jsonify({"message": f"Artículo con ID {item_id} no encontrado"}), 404

        # Crear el objeto JSON para la orden
        orden = {
            "datosRestaurante": {
                "restauranteId": restaurante_id,
                "nombreRestaurante": restaurante.get("nombreRestaurante")
                },
            "datosUsuario": {
                "usuarioId": usuario_id,
                "nombreUsuario": usuario.get("nombre"),
                "correoUsuario": usuario.get("correo")
                },
            "datosMenu": datos_menu,
            "estado": estado,
            "fechaOrden": fecha_orden,
            "montoTotal": total
        }

        # Insertar la orden en la base de datos
        db.Ordenes.insert_one(orden)
        return jsonify({"message": "Orden creada con éxito", "orden": orden}), 201
    except Exception as e:
        return jsonify({"message": "Error al crear la orden: " + str(e)}), 500


# Crear una reseña (Pedir campos por separado)
def crear_resena(request, db):
    try:
        # Pedir los campos por separado
        usuario_id_resena = request.json.get("usuarioId")
        restaurante_id_resena = request.json.get("restauranteId")
        comentario = request.json.get("comentario")
        calificacion = request.json.get("calificacion")
        fecha_resena = request.json.get("fechaResena")
        orden_id = request.json.get("ordenId")



        # Crear el objeto JSON para la reseña
        resena = {
            "usuarioId": usuario_id_resena,
            "restauranteId": restaurante_id_resena,
            "comentario": comentario,
            "calificacion": calificacion,
            "fechaResena": fecha_resena,
            "ordenId": orden_id
        }

        # Insertar la reseña en la colección 'resenas'
        result = db.Resenas.insert_one(resena)

        # Responder con un mensaje de éxito
        return jsonify({"message": f"Reseña creada con éxito, ID: {str(result.inserted_id)}"}), 201
    except Exception as e:
        # Manejo de errores
        return jsonify({"message": f"Error al crear la reseña: {str(e)}"}), 500
